### PR TITLE
Add 3D support

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #344
* #360

Enables PP+DP+TP and adds CI test case that runs on 8-gpu CI runner.

### CODE REVIEW

**Comment by wanchaol** [comment]
looks like 8GPU CI failed?

```
[rank0]:[rank0]:     all_local_plans, global_metadata = planner.create_global_plan(all_local_plans)
[rank0]:[rank0]:                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:[rank0]:   File "/opt/conda/envs/py_3.11/lib/python3.11/site-packages/torch/distributed/checkpoint/default_planner.py", line 121, in create_global_plan
[rank0]:[rank0]:     raise ValueError("Failed to validate global plan")
```

This looks like a DCP failure?

**Comment by fegin** [comment]
CI is still not happy even with https://github.com/pytorch/torchtitan/pull/360. Not sure what's going one, I could not reproduce the issue with the local machine.

**Comment by wconstab** [comment]
sorry for the noise here. the DCP failure is the one i fixed in the FSPD mesh PR.  I thought it would be included in last night's pytorch nightly 0604 since it landed yesterday, but it was landed too late in the day.

The 0605 nightly should fix this.  I have also updated this PR to include one more test, which reloads the saved PP checkpoint at step 10 and runs to step 20.  This validates the additional optimizer flattening logic that is in @fegin's #360 PR lower in this stack.

If CI passes tmrw then I propose to land both PRs.

File: .github/workflows/integration_test_8gpu.yaml [hunk 1]
```diff
@@ -0,0 +1,42 @@
+name: 8 GPU Integration Test
+
+on:
```
Comment (tianyu-l) [inline .github/workflows/integration_test_8gpu.yaml:3]: After chatted with Huy, I think this is OK. I can create another badge for the 8 GPU test after this PR lands.

**Review by wz337 — APPROVED** [review]

**Pull request merged**